### ISSUE TITLE
A very thick boundary is required to obtain accurate results

### ISSUE THREAD

**Issue by OXOLE1024** [IvoVellekoop/wavesim_py#1]
Hi,

I'm performing an acoustic wave equation simulation using the python version of WaveSim. I observed that an unusually thick boundary layer is required to obtain accurate results. Otherwise, the wave field at the boundary will become distorted.

### To Reproduce
Run code here to reproduce. 
```
import torch
import numpy as np
from tqdm import tqdm

from wavesim.helmholtzdomain import HelmholtzDomain
from wavesim.multidomain import MultiDomain
from wavesim.iteration import run_algorithm
from wavesim.utilities import preprocess

import matplotlib.pyplot as plt

class HelmholtzSolver:
    def __init__(self, omega, s, c0, dh, boundary_width):
        """
        Initialize the Helmholtz equation solver.

        Args:
            omega (float): Angular frequency (rad/s).
            s (np.ndarray): Complex slowness array, same shape as simulation domain.
            c0 (float): Reference speed.
            dh (float): Spatial step size (pixel size).
            boundary_width (int): Absorbing boundary width (in pixels).
        Returns:
            None. The solver is initialized for later use with the solve method.
        """
        ndim = s.ndim
        n = s * c0
        freq = omega / (2 * np.pi)
        wavelength = c0 / freq
        permittivity, _ = preprocess(n**2, boundary_width)
        
        periodic_flag = (False, False, True) if ndim == 2 else (False, False, False)
        domain = HelmholtzDomain(permittivity=permittivity,
                                 periodic=periodic_flag,
                                 pixel_size=dh,
                                 wavelength=wavelength,
                                 n_boundary=boundary_width)
        
        self.domain = domain
        self.shape = s.shape
        self.ndim = ndim
        self.wavelength = wavelength
        self.boundary_width = boundary_width
        
    def solve(self, src, max_iter=100000):
        """
        Solve the Helmholtz equation and return the wavefield solution.

        Args:
            src (np.ndarray): Source term, shape (batch_size, ...), matching the simulation domain dimensions.
            max_iter (int): Maximum number of iterations (default: 100000).
        Returns:
            u (np.ndarray): Wavefield solution, shape (batch_size, ...), same as input domain.
        """
        assert src.ndim == self.ndim+1, "Source dimension mismatch"
        batch_size = src.shape[0]

        src_amp = np.linalg.norm(src.reshape(batch_size, -1), axis=1)
        amp_scale = (2*np.pi/self.wavelength)**2 / src_amp
        src = src * amp_scale.reshape([batch_size] + [1] * (self.ndim))

        pad_width = ((0, 0),) + ((self.boundary_width, self.boundary_width),) * self.ndim
        src = np.pad(src, pad_width, mode='constant', constant_values=0)
        src = torch.tensor(src, dtype=torch.complex64)
        
        if self.ndim == 2:
            src = src.unsqueeze(-1)

        u = torch.zeros(batch_size, *self.shape, dtype=torch.complex64)
        for i in tqdm(range(batch_size), desc="Calculating wavefield of all sources"):
            u_computed, _, _ = run_algorithm(self.domain, src[i], max_iterations=max_iter)

            # Plot the wavefield with boundary layer
            plt.clf()
            plt.imshow(np.real(u_computed.squeeze().cpu().numpy()[:, :]), cmap='seismic', vmin=-0.03, vmax=0.03)
            plt.colorbar()
            plt.savefig(f"wvfield_bound.png", dpi=300)

            inner_slice = [slice(self.boundary_width, -self.boundary_width)] * self.ndim
            u[i] = u_computed.squeeze()[*inner_slice]
        u = u.cpu().numpy()
        u /= amp_scale.reshape([batch_size] + [1] * (self.ndim))
        
        return u

if __name__ == "__main__":
    freq = 5e5 # 500 kHz

    # Define a simple homogeneous medium
    sos = np.full((240, 240), 1500, dtype=np.float32)
    atten = np.full((240, 240), 0, dtype=np.float32)
    s = 1 / sos + 1j * atten / (2 * np.pi)

    c0 = 1500
    dh = 5e-4
    boundary_width = 32
    solver = HelmholtzSolver(freq*2*np.pi, s, c0, dh, boundary_width)

    # Set up a source near the boundary of the domain
    src = np.zeros((1, 240, 240), dtype=np.complex64)
    src[0, 220, 120] = 1e6 * (1 + 1j)

    u = solver.solve(src)

    # Plot the wavefield without boundary layer
    plt.clf()
    plt.imshow(np.real(u[0]), cmap='seismic', vmin=-0.05, vmax=0.05)
    plt.colorbar()
    plt.savefig(f"wvfield.png", dpi=300)
```

### Screenshots

![Image](https://github.com/user-attachments/assets/3674d943-83e6-4404-a3ae-a51ef82458a8)
*Wavefield snapshot showing boundary distortion (below the image) with a boundary width 32 pixels ~ 5 wavelengths*

![Image](https://github.com/user-attachments/assets/407dbbf3-6aeb-4842-b64f-9ccff27a35d6)
*Wavefield snapshot showing eliminated boundary distortion with an extremely thick boundary width 256 pixels ~ 43 wavelengths*

### Environment
- OS: Windows 10
- Python version: 3.12.7
- PyTorch version: 2.5.1
- CUDA version: 12.7

Thanks for your work on this solver — I'm happy to help debug further if needed!

**Comment by swapnilsmache** [comment]
Hi @OXOLE1024,
Thanks for your question! We have also observed this problem in some cases (also specifically in acoustic simulations) and are working on a solution.

And thank you for the MWE. I ran your simulation and compared the results with various boundary widths to the simulation with `boundary_wavelengths=100` (which I assumed as a "reference" without any wrapping artefacts), and found that `boundary_wavelengths>10` already gave a decent result (relative error of less than 0.1% with the "reference"). But you are right, at this point, the solution is to use thicker boundaries. 

Below I am adding some results for different boundary widths (11, 20, and 43 wavelengths). The 2D subplot shows the relative difference with the "reference" and the 1D cross sections along the x and y axis are taken at the source.

![Image](https://github.com/user-attachments/assets/d0b60fb9-3529-448f-ac99-2a39df9e122d)

![Image](https://github.com/user-attachments/assets/a9323370-c57c-4a6f-8c34-eb19f265ee65)

![Image](https://github.com/user-attachments/assets/6d5189de-5ed7-4149-a1fa-ff96b7d3d53d)

I added some plotting code to the code you shared. Attaching it below if you would like to rerun it.

```python
import torch
import numpy as np
from tqdm import tqdm

from wavesim.helmholtzdomain import HelmholtzDomain
from wavesim.multidomain import MultiDomain
from wavesim.iteration import run_algorithm
from wavesim.utilities import preprocess, relative_error

import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec

class HelmholtzSolver:
    def __init__(self, omega, s, c0, dh, boundary_width):
        """
        Initialize the Helmholtz equation solver.

        Args:
            omega (float): Angular frequency (rad/s).
            s (np.ndarray): Complex slowness array, same shape as simulation domain.
            c0 (float): Reference speed.
            dh (float): Spatial step size (pixel size).
            boundary_width (int): Absorbing boundary width (in pixels).
        Returns:
            None. The solver is initialized for later use with the solve method.
        """
        ndim = s.ndim
        n = s * c0
        freq = omega / (2 * np.pi)
        wavelength = c0 / freq
        permittivity, _ = preprocess(n**2, boundary_width)
        
        periodic_flag = (False, False, True) if ndim == 2 else (False, False, False)
        domain = HelmholtzDomain(permittivity=permittivity,
                                 periodic=periodic_flag,
                                 pixel_size=dh,
                                 wavelength=wavelength,
                                 n_boundary=boundary_width)
        
        self.domain = domain
        self.shape = s.shape
        self.ndim = ndim
        self.wavelength = wavelength
        self.boundary_width = boundary_width
        
    def solve(self, src, boundary_wavelengtths, max_iter=100000):
        """
        Solve the Helmholtz equation and return the wavefield solution.

        Args:
            src (np.ndarray): Source term, shape (batch_size, ...), matching the simulation domain dimensions.
            max_iter (int): Maximum number of iterations (default: 100000).
        Returns:
            u (np.ndarray): Wavefield solution, shape (batch_size, ...), same as input domain.
        """
        assert src.ndim == self.ndim+1, "Source dimension mismatch"
        batch_size = src.shape[0]

        src_amp = np.linalg.norm(src.reshape(batch_size, -1), axis=1)
        amp_scale = (2*np.pi/self.wavelength)**2 / src_amp
        src = src * amp_scale.reshape([batch_size] + [1] * (self.ndim))

        pad_width = ((0, 0),) + ((self.boundary_width, self.boundary_width),) * self.ndim
        src = np.pad(src, pad_width, mode='constant', constant_values=0)
        src = torch.tensor(src, dtype=torch.complex64)
        
        if self.ndim == 2:
            src = src.unsqueeze(-1)

        u = np.zeros((batch_size, *self.shape), dtype=np.complex64)
        for i in tqdm(range(batch_size), desc="Calculating wavefield of all sources"):
            u_computed = run_algorithm(self.domain, src[i], max_iterations=max_iter)[0].squeeze().cpu().numpy()

            # Plot the wavefield with boundary layer
            plt.clf()
            plt.imshow(np.real(u_computed), cmap='seismic', vmin=-0.03, vmax=0.03)
            plt.colorbar()
            plt.savefig(f"wvfield_bound_bw{boundary_wavelengtths}.png", dpi=300)

            # Plot the wavefield with boundary layer (+ 1d cross sections)
            fig = plt.figure(figsize=(10, 10))
            gs = GridSpec(2, 2, figure=fig, height_ratios=[3, 1], width_ratios=[1, 3], hspace=0.1, wspace=0.1)

            ax1 = fig.add_subplot(gs[0, 0])
            ax1.plot(np.real(u_computed[:, u_computed.shape[1]//2]), np.arange(u_computed.shape[0]))
            ax1.set_title("Cross section at source")
            ax1.set_ylabel("y")
            ax1.set_ylim(0, u_computed.shape[0])
            ax1.invert_yaxis()

            ax2 = fig.add_subplot(gs[0, 1])
            ax2.imshow(np.real(u_computed[:, :]), cmap='seismic', vmin=-0.03, vmax=0.03, extent=[0, u_computed.shape[1], u_computed.shape[0], 0])
            ax2.set_title("Wavefield with boundary layer")
            ax2.set_xticks([])
            ax2.set_yticks([])
            ax2.hlines(220 + self.boundary_width, 0, u_computed.shape[1], color='k', lw=1, ls='-.')
            ax2.vlines(u_computed.shape[1]//2, 0, u_computed.shape[0], color='k', lw=1, ls='-.')

            ax3 = fig.add_subplot(gs[1, 1])
            ax3.plot(np.real(u_computed[220 + self.boundary_width, :]))
            ax3.set_title("Cross section at source")
            ax3.set_xlabel("x")
            ax3.set_xlim(0, u_computed.shape[1])

            plt.savefig(f"wvfield_bound_cs_bw{boundary_wavelengtths}.png", dpi=300, bbox_inches='tight', pad_inches=0.1)
            plt.close(fig)

            inner_slice = [slice(self.boundary_width, -self.boundary_width)] * self.ndim
            u[i] = u_computed[*inner_slice]
        u /= amp_scale.reshape([batch_size] + [1] * (self.ndim))
        
        return u

if __name__ == "__main__":
    freq = 5e5 # 500 kHz

    # Define a simple homogeneous medium
    sos = np.full((240, 240), 1500, dtype=np.float32)
    atten = np.full((240, 240), 0, dtype=np.float32)
    s = 1 / sos + 1j * atten / (2 * np.pi)

    c0 = 1500
    dh = 5e-4

    wavelength = c0 / freq
    boundary_wavelengths = 11
    boundary_width = int(boundary_wavelengths * wavelength / dh)
    print(f"Boundary width: {boundary_width} pixels")
    # boundary_width = 256
    solver = HelmholtzSolver(freq*2*np.pi, s, c0, dh, boundary_width)

    # Set up a source near the boundary of the domain
    src = np.zeros((1, 240, 240), dtype=np.complex64)
    src[0, 220, 120] = 1e6 * (1 + 1j)

    u = solver.solve(src, boundary_wavelengths).squeeze()
    # np.save(f"wavefield_bw{boundary_wavelengths}.npy", u)
    u_ref = np.load(f"wavefield_bw100.npy")
    print(f"Relative error: {relative_error(u, u_ref):.2e}")

    # Plot the wavefield without boundary layer
    plt.clf()
    plt.imshow(np.real(u), cmap='seismic', vmin=-0.05, vmax=0.05)
    plt.colorbar()
    plt.savefig(f"wvfield_bw{boundary_wavelengths}.png", dpi=300)
    plt.close()

    # Plot the wavefield without boundary layer (+ 1d cross sections at source location)
    fig = plt.figure(figsize=(10, 10))
    gs = GridSpec(2, 2, figure=fig, height_ratios=[3, 1], width_ratios=[1, 3], hspace=0.1, wspace=0.1)

    ax1 = fig.add_subplot(gs[0, 0])
    ax1.plot(np.real(u[:, u.shape[1]//2]), np.arange(u.shape[0]))
    ax1.set_title("Cross section at source")
    ax1.set_ylabel("y")
    ax1.set_ylim(0, u.shape[0])
    ax1.invert_yaxis()

    ax2 = fig.add_subplot(gs[0, 1])
    ax2.imshow(np.real(u[:, :]), cmap='seismic', vmin=-0.03, vmax=0.03, extent=[0, u.shape[1], u.shape[0], 0])
    ax2.set_title("Wavefield without boundary layer")
    ax2.set_xticks([])
    ax2.set_yticks([])
    ax2.hlines(220, 0, u.shape[1], color='k', lw=1, ls='-.')
    ax2.vlines(u.shape[1]//2, 0, u.shape[0], color='k', lw=1, ls='-.')

    ax3 = fig.add_subplot(gs[1, 1])
    ax3.plot(np.real(u[220, :]))
    ax3.set_title("Cross section at source")
    ax3.set_xlabel("x")
    ax3.set_xlim(0, u.shape[1])

    plt.savefig(f"wvfield_cs_bw{boundary_wavelengths}.png", dpi=300, bbox_inches='tight', pad_inches=0.1)
    plt.close(fig)

    # Plot the difference between the two wavefields (+ 1d cross sections at source location)
    u_diff = (u - u_ref) / u_ref

    fig = plt.figure(figsize=(10, 10))
    gs = GridSpec(2, 3, figure=fig, height_ratios=[3, 1], width_ratios=[1, 3.4, 0.1], hspace=0.1, wspace=0.1)

    ax1 = fig.add_subplot(gs[0, 0])
    ax1.plot(np.real(u_diff[:, u_diff.shape[1]//2]), np.arange(u_diff.shape[0]))
    ax1.set_title("Cross section at source")
    ax1.set_ylabel("y")
    ax1.set_ylim(0, u_diff.shape[0])
    ax1.set_xlim(-1, 1)
    ax1.invert_yaxis()

    ax2 = fig.add_subplot(gs[0, 1])
    im2 = ax2.imshow(np.real(u_diff[:, :]), cmap='seismic', vmin=-1, vmax=1, extent=[0, u_diff.shape[1], u_diff.shape[0], 0])
    ax2.set_title(fr"Wavefield relative difference (bw={boundary_wavelengths}$\lambda$={boundary_width} pixels) (RE {relative_error(u, u_ref):.2e})")
    ax2.set_xticks([])
    ax2.set_yticks([])
    ax2.hlines(220, 0, u_diff.shape[1], color='k', lw=1, ls='-.')
    ax2.vlines(u_diff.shape[1]//2, 0, u_diff.shape[0], color='k', lw=1, ls='-.')
    
    ax3 = fig.add_subplot(gs[0, 2])
    ax3.set_axis_off()
    plt.colorbar(mappable=im2, ax=ax3, fraction=1.5, aspect=30)

    ax3 = fig.add_subplot(gs[1, 1])
    ax3.plot(np.real(u_diff[220, :]))
    ax3.set_title("Cross section at source")
    ax3.set_xlabel("x")
    ax3.set_xlim(0, u_diff.shape[1])
    ax3.set_ylim(-1, 1)

    plt.savefig(f"wvfield_diff_cs_bw{boundary_wavelengths}.png", dpi=300, bbox_inches='tight', pad_inches=0.1)
    plt.close(fig)
```
